### PR TITLE
Add available images tile to Devices

### PR DESCRIPTION
**Pull request by boaz0**
- [x] unit tests
- [x] create tiles component and available images tile
- [x] add tiles to Devices
- [x] create images store
- [x] create loadImages action and api
- [x] on failure display notification toast and display message saying "No images to view"

### CODE REVIEW

File: src/components/Tiles.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  const { getRegistry } = useContext(RegistryContext);
+  const dispatch = useDispatch();
+  useEffect(() => {
+    const registered = getRegistry().register({
+      imagesReducer,
+      notifications: notificationsReducer,
+    });
+    loadImages(dispatch);
+    return () => registered();
+  }, [dispatch, loadImages]);
```
Comment (karelhala) [inline src/components/Tiles.js:20]: No need to add `loadImages` to the dependency list.

File: src/components/AvailableImagesTile.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+AvailableImageTileBase.propTypes = {
+  children: PropTypes.children,
+};
+
+const AvailableImageTile = () => {
+  const { status, data } = useSelector(({ imagesReducer }) => ({
+    status: imagesReducer?.status || 'PENDING',
+    data: imagesReducer?.data || null,
+    error: imagesReducer?.error || null,
+  }));
```
Comment (karelhala) [inline src/components/AvailableImagesTile.js:30]: Either spread this to 2 selectors or use `shallowEqual` compare function so we don't get multiple re-renders.

File: src/store/images.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+const loadImagesFulfilled = (state, { payload }) => ({
+  ...state,
+  data: payload,
+  status: 'DONE',
+});
+
+export default applyReducerHash(
+  {
+    [`${LOAD_IMAGES}_PENDING`]: loadImagesPending,
+    [`${LOAD_IMAGES}_FULFILLED`]: loadImagesFulfilled,
```
Comment (karelhala) [inline src/store/images.js:20]: Error state can be caught here
```suggestion
    [`${LOAD_IMAGES}_FULFILLED`]: loadImagesFulfilled,
    [`${LOAD_IMAGES}_ERROR`]: () => ({ loaded: true, hasError: true }),
```

File: src/components/AvailableImagesTile.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  }
+  if (status === 'DONE' && data.success) {
+    content = (
+      <>
+        <CardBody>
+          <Button variant="link" style={{ paddingLeft: 0 }}>
+            {data.payload.meta.count} images
+          </Button>
+        </CardBody>
+        <CardBody>
```
Comment (karelhala) [inline src/components/AvailableImagesTile.js:60]: I would use `CardFooter` here. There's something strange about two bodies :smiley:

File: src/components/AvailableImagesTile.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  if (status === 'PENDING') {
+    content = (
+      <>
+        <CardBody
+          isFilled
+          style={{
+            display: 'flex',
+            justifyContent: 'center',
+            alignItems: 'center',
+          }}
```
Comment (karelhala) [inline src/components/AvailableImagesTile.js:42]: This can be resolved trough `Bullseye` layout.

File: src/components/AvailableImagesTile.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        </CardBody>
+      </>
+    );
+  }
+  if (status === 'DONE' && !data.success) {
+    content = <CardBody>{data.payload}</CardBody>;
+  }
+  if (status === 'DONE' && data.success) {
+    content = (
+      <>
```
Comment (karelhala) [inline src/components/AvailableImagesTile.js:54]: I don't really like this, it's easy to overlook and you have to think what it actually does. Much cleaner approach is with `Fragment` component from React.

File: src/components/AvailableImagesTile.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+            alignItems: 'center',
+          }}
+        >
+          <Spinner />;
+        </CardBody>
+      </>
+    );
+  }
+  if (status === 'DONE' && !data.success) {
+    content = <CardBody>{data.payload}</CardBody>;
```
Comment (karelhala) [inline src/components/AvailableImagesTile.js:50]: No need to assign everytime the card body to `content`. Much smoother apoprach is trough directly returning the value
```JSX
return <AvailableImageTileBase><CardBody>{data.payload}</CardBody></AvailableImageTileBase>;
```

You can then directly remove the `<>`/`<Fragment>` component

File: src/components/AvailableImagesTile.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+import { useSelector } from 'react-redux';
+
+const AvailableImageTileBase = ({ children }) => {
+  return (
+    <Card>
+      <CardTitle>Available images</CardTitle>
+      {children}
+    </Card>
+  );
+};
```
Comment (karelhala) [inline src/components/AvailableImagesTile.js:19]: Extra return statement
```suggestion
const AvailableImageTileBase = ({ children }) => (
  <Card>
    <CardTitle>Available images</CardTitle>
    {children}
  </Card>
);
```

**Review by karelhala — COMMENTED** [review]
Looking good! I am not a big fan of string constants, we shoud stick to the `loaded: true` so we don't mix two formats.

File: src/store/images.js [hunk 9]
```diff
@@ -0,0 +1,23 @@
+import { applyReducerHash } from '@redhat-cloud-services/frontend-components-utilities/ReducerRegistry';
+import { LOAD_IMAGES } from './action-types';
+
+const initialState = {};
+
+const loadImagesPending = (state) => ({
+  ...state,
+  data: null,
+  status: 'PENDING',
```
[the remark this answers is not in the document]
Comment (boaz0) [reply src/store/images.js:9]: I changed it to `isLoading` because I saw that's how you did it with Devices and Groups.

File: src/api/index.js [hunk 10]
```diff
@@ -1,4 +1,5 @@
 import { statusMapper } from '../constants';
+import axios from 'axios';
```
Comment (karelhala) [inline src/api/index.js:2]: It should be better to use `instance` from frontend-components package, so we have all the interceptors.
```suggestion
import { instance } from '@redhat-cloud-services/frontend-components-utilities/interceptors/interceptors';
```

File: src/api/index.js [hunk 11]
```diff
@@ -135,3 +136,7 @@ export const updateGroup = ({ uuid, systemIDs, groupName }) => {
   }
   return Promise.resolve();
 };
+
+export const fetchActiveImages = () => {
+  return axios('/api/image-builder/v1/composes');
```
Comment (karelhala) [inline src/api/index.js:141]: You can then call `instance.get`.

File: src/store/actions.js [hunk 12]
```diff
@@ -67,3 +69,22 @@ export const preSelectEntity = (id, selected) => ({
...
+          variant: 'danger',
+          title: 'Can not show images data',
+          description: 'Failed receiving images from image-builder',
+        },
+      },
+    },
+    // the '.catch' part is necessary because redux-promise-middleware throws the error on REJECTED
+    // and to avoid the app exploding I need to catch it here.
+    // THANK you redux-promise-middleware for not allowing to customize this behavior. 😠
+  }).catch(() => null);
```
Comment (karelhala) [inline src/store/actions.js:89]: Doesn't this prevent the promise-middleware from firing `REJECTED` state? What exactly is the problem? An error in console or completely broken UI?

File: src/store/images.js [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+
+const loadActiveImagesPending = () => ({
+  isLoading: true,
+  hasError: false,
+  data: null,
+});
+
+const loadActiveImagesFulfilled = (_, { payload }) => {
+  return {
+    isLoading: false,
```
Comment (karelhala) [inline src/store/images.js:14]: You might want to spread the previous state in here as well.

**Review by karelhala — COMMENTED** [review]
Looking good! Just a few minor nitpicks, but overall looks solid!

File: src/store/actions.js (hunk 12 above)
Comment (boaz0) [reply src/store/actions.js:89]: completely breaking UI.

If you look here: https://github.com/pburtchaell/redux-promise-middleware/blob/main/src/index.js#L162-L165
On Promise reject it dispatched REJECTED action and a line after `throw` the error again.
Looking in their examples, one way to deal with it is catching it in the action: 
https://github.com/pburtchaell/redux-promise-middleware/blob/main/examples/catching-errors/index.js#L19-L27

**Review by karelhala — APPROVED** [review]

**Pull request merged**